### PR TITLE
OP-1020 | DELETE last movement in Pharmaceutical Stock

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OP-1020.

### CODE REVIEW

File: doc_user/UserManual.adoc [hunk 1]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
 movement, several discharging movements may be generated. Ask the Administrator or check the _Administrator’s Guide_
 for more information.
 
-<<<
+[#delete-last-stock-movement]
+===== 4.2.2.3 Delete Last Stock Movement (*[.underline]##D##elete*)
+To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. if it is not the last movement you will have and *_Warning_* window:
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]:
```suggestion
To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. If it is not the last movement you will see a *_Warning_* window:
```

File: doc_user/UserManual.adoc [hunk 2]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+
+image:image250.png[Warning]
+
+But, if the row selected is the last movement you will have the *_Question_* window:
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]:
```suggestion
But if the row selected is the last movement the *_Question_* window is presented:
```

File: doc_user/UserManual.adoc [hunk 3]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+
+image:image248.png[Question]
+
+Click *[.underline]##Y##es* if you realy want to delete the last movement or click *[.underline]##N##o* if no. Before click the *[.underline]##Y##es* button to be sure that the product present in the last movement is not used in the *_Pharmaceuticals Stock [.underline]##W##ard_*
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]:
```suggestion
Click *[.underline]##Y##es* to confirm the deletion of the last movement or click *[.underline]##N##o* to not  delete the movement. Before clicking the *[.underline]##Y##es* button check that the product present in the last movement is not used in the *_Pharmaceuticals Stock [.underline]##W##ard_*
```

File: doc_user/UserManual.adoc [hunk 4]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+if yes you will have the *_Error_* window.
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]:
```suggestion
if it is present then the *_Error_* window is shown.
```

File: doc_user/UserManual.adoc [hunk 5]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+
+image:image249.png[Error]
+
+If no, the last stock movement will be deleted from the table and you will have the *_Success_* window:
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]:
```suggestion
If it is not present then the last stock movement is deleted from the table and the *_Success_* window is shown:
```

File: doc_user/UserManual.adoc (hunk 3 above)
Comment (dbmalkovsky) [reply doc_user/UserManual.adoc]: The documentation tries to avoid at all cost the use of *you*

File: doc_user/UserManual.adoc [hunk 6]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+
+image:image251.png[success]
+
+NOTE: This button can be deactivated if you don't want to allow movements to be deleted. Ask the Administrator or check the Administrator's Guide for more information.
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]:
```suggestion
NOTE: This button can be deactivated to not allow movements to be deleted. Ask the Administrator or check the Administrator's Guide for more information.
```

File: doc_user/UserManual.adoc [hunk 7]
```diff
@@ -630,7 +631,26 @@ to the expiring date. If the first selected lot does not contain a high enough q
 movement, several discharging movements may be generated. Ask the Administrator or check the _Administrator’s Guide_
 for more information.
 
-<<<
+[#delete-last-stock-movement]
+===== 4.2.2.3 Delete Last Stock Movement (*[.underline]##D##elete*)
+To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. if it is not the last movement you will have and *_Warning_* window:
+
+image:image250.png[Warning]
```
Comment (dbmalkovsky) [inline doc_user/UserManual.adoc]: This should be an ERROR and not just a warning; see dialog title.

Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: the title is Warning but should I rename it to error ?

Comment (dbmalkovsky) [reply doc_user/UserManual.adoc]: I think so as a warning to me means something to know about but we will continue anyway.   Doesn't this really mean you can't do it?  So it should be IMHO an error.

Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: okay you right but we already merge the gui so can I create the task to update the window and use error dialog instead of warning dialog?

Comment (mwithi) [reply doc_user/UserManual.adoc]: It is actually just warning even before the operation takes place, so I think it is different from an ERROR which comes during an operation, but I think the point here is consistency with the rest of the doc.

According with the PR merged in GUI it is a Warning, maybe we can have it consistent with the GUI as for now and change it later on?

File: doc_user/UserManual.adoc [hunk 8]
```diff
@@ -491,6 +491,7 @@ The following functions are available from the *Buttons panel* of the window *_S
 
 * *C[.underline]##h##arge*: record a new charge movement.
 * *[.underline]##D##ischarge*: record a new discharge movement.
+* *[.underline]##D##elete*: delete the last movement.
```
Comment (mwithi) [inline doc_user/UserManual.adoc]: The underlined character should be the one used as _Mnemonic_ in the code and cannot be shared with other buttons in the same window.

According with the current code in develop now, the Delete button has been mapped with common `angal.common.delete.btn.key=D`, but in this windows we might map differently because we already have `angal.medicalstock.discharge.btn.key=D`, so we could define new `angal.medicalstock.discharge.btn=Delete` and `angal.medicalstock.delete.btn.key=L` or having `angal.medicalstock.discharge.btn.key=R`.

Since we did the same for Charge and Close button (keeping the "C" for common operation "Close") I would assume the best choice is to keep "D" for common operation of "Delete" and having Charge=H and Discharge=R.

In any case this doc should be updated accordingly.

File: doc_user/UserManual.adoc (hunk 7 above)
Comment (mwithi) [reply doc_user/UserManual.adoc]: Anyway, `image250` it is showing only if no movement is selected.

The correct dialog shown when the user select a "not-last movement", it's still a warning but like this one:

image

File: doc_user/UserManual.adoc (hunk 8 above)
Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: please should I apply those changes now so that after you will create the task to update it on gui ?

Comment (mwithi) [reply doc_user/UserManual.adoc]: we are still in the matter of OP-1020, just create a new branch from the latest develop and open a new PR for it

Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: and this branch must contain what I don't understand please.

Comment (mwithi) [reply doc_user/UserManual.adoc]: > I would assume the best choice is to keep "D" for common operation of "Delete" and having Charge=H and Discharge=R.

this

File: doc_user/UserManual.adoc (hunk 7 above)
Comment (mwithi) [reply doc_user/UserManual.adoc]: I've checked, and according with the rest of the code and documentation, we used warning only few times and for not-blocking operations, we used error for all other cases instead.

So @ArnaudFonzam you can include this change in the new PR on GUI, thanks!

File: doc_user/UserManual.adoc (hunk 8 above)
Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: Okay the task is to just change the  key of charge and discharge button

File: doc_user/UserManual.adoc (hunk 7 above)
Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: Okay

Comment (mwithi) [reply doc_user/UserManual.adoc]: and update this doc accordingly, thanks!

Comment (ArnaudFonzam) [reply doc_user/UserManual.adoc]: okay I'm working on it

File: doc_user/UserManual.adoc [hunk 9]
```diff
@@ -630,7 +631,32 @@ to the expiring date. If the first selected lot does not contain a high enough q
 movement, several discharging movements may be generated. Ask the Administrator or check the _Administrator’s Guide_
 for more information.
 
-<<<
```
Comment (mwithi) [inline doc_user/UserManual.adoc:633]: revert this deletion, this symbol is needed for the PDF pagination

File: doc_user/UserManual.adoc [hunk 10]
```diff
@@ -630,7 +631,32 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+
+image:image250.png[Error]
+
+if you don't choosed any row and:
+
+image:image252.png[Error]
+
+if you choosed more than one row.
+
+But if the row selected is the last movement the *_Question_* window is presented:
```
Comment (mwithi) [inline doc_user/UserManual.adoc]: Let's simplify this part. First of all we have to remove the NOTE in line 500. Then, as already suggested, we remove all "you" occurences and make it all impersonal. Then I would rephrase it as follow, and avoid to explain all possible errors. 

Please upload the _missing image_.

```suggestion
In Pharmaceutical Stock it is not possible to change the past and all operations need to be done in consecutive order. If a mistake concern a movement in the past (for example, the wrong quantity of a pharmaceutical is inserted in charge), then use a new opposite movement to correct the mistake (in the example, use a discharge movement to undo the unwanted change).

Nevertheless, the admin can enable/disable the *[.underline]##D##elete* button that allows the enabled user do delete only the last movement, and helps quickly undoing mistakes.

To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. Only one row is allowed to delete at a time.

If it is not the last movement the following  *_Error_* message is shown:

image:<missing image>[Not last movement error]

But if the row selected is the last movement this *_Question_* message is presented:
```

File: doc_user/UserManual.adoc [hunk 11]
```diff
@@ -630,7 +631,32 @@ to the expiring date. If the first selected lot does not contain a high enough q
...
+Click *[.underline]##Y##es* if you want to confirm the deletion of the last movement or click *[.underline]##N##o* if you do not want to delete the movement. Before clicking the *[.underline]##Y##es* button check that the product present in the last movement is not used in the *_Pharmaceuticals Stock [.underline]##W##ard_*
+if it is then the *_Error_* window is shown.
+
+image:image249.png[Error]
+
+If it is not present then the last stock movement will be deleted from the table and the *_Success_* window is shown:
+
+image:image251.png[success]
+
+NOTE: This button can be deactivated to not allow movements to be deleted. Ask the Administrator or check the Administrator's Guide for more information.
```
Comment (mwithi) [inline doc_user/UserManual.adoc]:
```suggestion
Click *[.underline]##Y##es* to confirm the deletion of the last movement, otherwise click *[.underline]##N##o* to cancel the operation. After clicking the *[.underline]##Y##es* button, the system checks that the product involved the selected movement has not been used itself in the *_Pharmaceuticals Stock [.underline]##W##ard_*.

If the product has been already used in a Ward, then the *_Error_* message is shown.

image:image249.png[Error]

Otherwise the selected movement is deleted and the *_Success_* window is shown:

image:image251.png[success]

NOTE: This button can be enabled/disabled to allow/not allow movements deletion. Ask the Administrator or check the Administrator's Guide for more information.
```

File: doc_user/UserManual.adoc [hunk 12]
```diff
@@ -632,6 +633,34 @@ for more information.
...
+
+If it is not the last movement the following  *_Error_* message is shown:
+
+image:image250.png[Not last movement error]
+
+But if the row selected is the last movement this *_Question_* message is presented:
+
+image:image248.png[Question]
+
+Click *[.underline]##Y##es* to confirm the deletion of the last movement, otherwise click *[.underline]##N##o* to cancel the operation. After clicking the *[.underline]##Y##es* button, the system checks that the product involved the selected movement has not been used itself in the *_Pharmaceuticals Stock [.underline]##W##ard_*.
```
Comment (mwithi) [inline doc_user/UserManual.adoc]:
```suggestion
Click *[.underline]##Y##es* to confirm the deletion of the last movement, otherwise click *[.underline]##N##o* to cancel the operation. After clicking the *[.underline]##Y##es* button, the system checks if the product involved in the selected movement has not been used itself in the *_Pharmaceuticals Stock [.underline]##W##ard_* (see <<pharmaceutical-stock-ward,Pharmaceuticals Stock Ward>> in this document).
```

File: doc_user/UserManual.adoc [hunk 13]
```diff
@@ -632,6 +633,34 @@ for more information.
 
 <<<
 
+[#delete-last-stock-movement]
+===== 4.2.2.3 Delete Last Stock Movement (*[.underline]##D##elete*)
+In Pharmaceutical Stock it is not possible to change the past and all operations need to be done in consecutive order. If a mistake concern a movement in the past (for example, the wrong quantity of a pharmaceutical is inserted in charge), then use a new opposite movement to correct the mistake (in the example, use a discharge movement to undo the unwanted change).
```
Comment (mwithi) [inline doc_user/UserManual.adoc]:
```suggestion
In Pharmaceutical Stock it is not possible to change the past and all operations need to be done in consecutive order. If a mistake concerns a movement in the past (for example, the wrong quantity of a pharmaceutical is inserted in charge), then use a new opposite movement to correct the mistake (in the example, use a discharge movement to undo the unwanted change).
```

File: doc_user/UserManual.adoc [hunk 14]
```diff
@@ -632,6 +633,34 @@ for more information.
...
+
+Nevertheless, the admin can enable/disable the *[.underline]##D##elete* button that allows the enabled user do delete only the last movement, and helps quickly undoing mistakes.
+
+To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. Only one row is allowed to delete at a time.
+
+If it is not the last movement the following  *_Error_* message is shown:
+
+image:image250.png[Not last movement error]
```
Comment (mwithi) [inline doc_user/UserManual.adoc:644]: Please this image is still the wrong one

File: doc_user/UserManual.adoc
Comment (mwithi) [inline doc_user/UserManual.adoc:1]: The NOTE at line 500 is still there, please remove it.

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!
